Adding loader for Filosax

dave-foster opened the pull request:
All checkboxes checked!

dave-foster commented:
Hi mirdata team,

I'm struggling to pass some of the tests, despite my best efforts… Firstly, there is failure in `buildpy36/37/38` due to the lack of exception when returning a `cached_property` (a wrapped dictionary of note properties), for which I've tried to fix with different permutations of exceptions. Secondly, there is an error in the docs: `AttributeError: 'Values' object has no attribute 'section_self_link'` for which I've been unable to rectify. Would it be possible to advise the best approaches to these problems? Also, if it were possible to have this looked at pre-ISMIR, I would be eternally grateful!

Cheers,

Dave

rabitt commented:
>> Hi mirdata team,
> 
> I'm struggling to pass some of the tests, despite my best efforts… Firstly, there is failure in buildpy36/37/38 due to the lack of exception when returning a cached_property (a wrapped dictionary of note properties), for which I've tried to fix with different permutations of exceptions. Secondly, there is an error in the docs: AttributeError: 'Values' object has no attribute 'section_self_link' for which I've been unable to rectify. Would it be possible to advise the best approaches to these problems? Also, if it were possible to have this looked at pre-ISMIR, I would be eternally grateful!
> 
> Cheers,
> 
> Dave

Hi Dave, thanks for the PR! We'll do our best to take a look in the next day or two, to have it ready by the end of the ISMIR tutorials.

rabitt commented:
@dave-foster quick question - in the download info, you mention a script to run:
```
python Scripts/Compile_Backing.py -version full
```
I had a quick look at the conversation in the issues about this - where does this script live now? One option is to point to e.g. another github repo of yours with the script. Another is to create a subfolder in mirdata, e.g.  `mirdata/scripts/preprocessing` and create a `preprocess_filosax.py` script in there. What do you prefer?

dave-foster commented:
Hi Rachel, thanks for looking at this - I'm reminded of what a bummer it is that we aren't all making our way to Bangalore to be able to chat in person!

The script that you mention is in the material that users download from the Zenodo site (see a couple of lines earlier for the "Lite" link). This works fine as a pre-processing step, so could stay like that, or be incorporated into mirdata, whichever works best. Should I upload the script somewhere to see what you reckon?

I would point out that if any of this gives the impression that I have the faintest idea what I'm doing, then please don't be deceived!

dave-foster commented:
Hi @rabitt et al, a friendly bump on this PR, to see if anybody could help me out in getting it approved? I've returned to the code this week to make some updates, and it turns out that I still haven't acquired the skills to get it through the unit tests!

nkundiushuti commented:
hi! could we help you merge this? I think it currently fails.

dave-foster commented:
Thanks Marius, I'd kind of given up hope of it ever getting merged, so any help you could give would be greatly appreciated!

nkundiushuti commented:
if you could try another commit and push then I could check. I am unable to see the errors at the moment

dave-foster commented:
Thanks Marius, I've re-committed and generated the failing test logs again. Hopefully you can see them now? From what I can make out, the problem seems to be with the code for accessing the note annotations. These are available for each of the sax tracks, but not the backing track. It is complaining because an exception is raised when the user tries to access annotations on a backing track - but then it also complains (I seem to remember) when an exception isn't raised (and it fails gracefully)! I've tried various re-formatting of this code to try and get it to conform.

nkundiushuti commented:
there are two errors here:
- one unrelated with this dataloader. I can look into that.
- one related with this dataloader concerning the notes property:
```
    @core.cached_property
    def notes(self) -> Optional[List[Note]]:
        """The track's note list - only for Sax files
    
        Returns:
            * [Note] - ordered list of Note objects
    
        """
        note_path = "" if self.annotation_path == None else self.annotation_path
        try:
            with open(note_path) as fhandle:
                note_dict = json.load(fhandle)["notes"]
        except FileNotFoundError:
>           raise FileNotFoundError("Note data only available for Sax tracks.")
E           FileNotFoundError: Note data only available for Sax tracks.

mirdata/datasets/filosax.py:257: FileNotFoundError
```
let me fix the first and then we will get back to you

dave-foster commented:
Thanks Marius, I'll stand by and await further instruction. The issue with the notes property is what has been thwarting the PR since the first attempt in November 2021 (see my initial comment)! The code originally failed gracefully when the user attempted to access the `Notes` annotations on a backing track, but the circleci tests required an exception to be raised. When I raise an exception, of any type/style, this also causes the circleci tests to fail. I'm sure that there is a different way of coding it so that it works - I was darned if I could find it, though…

dave-foster commented:
Thanks Marius, I've updated with the master. Let me know if you have any suggestions on how to reformat to avoid the exception/no-exception problem.

nkundiushuti commented:
hi David. sorry I have been busy and I couldn't look into this. looking at the code could you check if the track is not a backing track and then return the notes otherwise return an empty dict/list?

dave-foster commented:
Thanks Marius, that's what I was doing in the first place. I've implemented it again - as before, it now complains because it's not raising an exception! Any other ideas?

nkundiushuti commented:
hi Dave!
there is an easy fix to that.
at the moment you are doing
```
        if note_path == "":
             print("Note data only available for Sax tracks.")
             note_dict = {}
         else:
             with open(note_path) as fhandle:
                 note_dict = json.load(fhandle)["notes"]
         return [Note(n) for n in note_dict]
```

I propose you return an empty note list or create list with a single element, a dummy empty note 
```
        if note_path == "":
             print("Note data only available for Sax tracks.")
             note_dict = {}
             return [Note(n) for n in note_dict]
         else:
             with open(note_path) as fhandle:
                 note_dict = json.load(fhandle)["notes"]
         return [Note(n) for n in note_dict]
```

dave-foster commented:
Thanks Marius - I'm afraid that your code doesn't work either. Returning an empty list, or one with a single dummy Note (or anything), still gives an error that it's expecting an Exception to be raised. Raising any kind of Exception also gives an error. Any other ideas?

genisplaja commented:
Hello @dave-foster! I think you are forgetting to indicate, in the `tests/test_loaders.py` the testing track. You should edit this file, and add your test file in `CUSTOM_TEST_TRACKS` as `"filosax": "multitrack_02_sax_1"`. That should help.
Also, you might want to define `notes` in the Track class like this:
```
@core.cached_property
def notes(self) -> List[Note]:
   """The track's note list - only for Sax files

   Returns:
       * [Note] - ordered list of Note objects

   """
   return load_notes(self.annotation_path)
```
and then have a function to load the notes like that:
```
@io.coerce_to_string_io
def load_notes(fhandle: BinaryIO) -> List[Note]:
    """docs here

    Returns:
        * List[Notes]

    """
    if fhandle:
        note_dict = json.load(fhandle)["notes"]
        return [Note(n) for n in note_dict]
    else:
        return [Note[{"a_start_time": 0.0}]]
```
or maybe unifying this function with the `load_annotations` function you do have below. 

Hope this helps!!

dave-foster commented:
By Jove, he's got it! Brilliant, thanks @genisplaja. I had no idea I needed to add the info to the `tests/test_loaders.py` file.

I'm left now with a fail on the `codecov/patch` test, where it looks like it's complaining that the annotations for the backing track files (bars, chords, track name, etc.) aren't being tested. (I think they are tested in the `test_to_jams` function though) Can you add 2 tracks to the `test_tracks` function? If I change that from a sax track to a backing track, it won't then test the sax-specific annotations (and will presumable complain about that instead)!

genisplaja commented:
Great, sorry for the delay on finding that out :)

See the following example: [https://github.com/mir-dataset-loaders/mirdata/blob/c4759b4c75ab9a3b364cf9ccb62b7adb65cf2c10/tests/datasets/test_saraga_carnatic.py#L475](https://github.com/mir-dataset-loaders/mirdata/blob/c4759b4c75ab9a3b364cf9ccb62b7adb65cf2c10/tests/datasets/test_saraga_carnatic.py#L475)
In this case, an additional testing example is loaded here to test the case when a particular annotation is not available for certain tracks. You can just leave the default testing track in `test_loaders.py` as it is now, and momentarily load an additional one to cover the untested lines. No need to include the additional track in `test_loaders.py`. If I'm not mistaken, it should work as the example I attached. 

Let me know if you have further questions!

genisplaja commented on mirdata/datasets/filosax.py:
```diff
@@ -0,0 +1,512 @@
+        pdf_path (str): path to PDF file
+
+    Cached Properties:
+        notes ([Note]): an ordered list of Note objects
```
Could you please write that as `(list, Note)`? Yours is not incorrect :) Just to be consistent with the rest of the docs.

genisplaja commented on mirdata/datasets/filosax.py:
```diff
@@ -0,0 +1,512 @@
+            returns the audio to be mixed
+        name (str): the name of the tune
+        duration (float): the duration, in seconds
+        beats ([Observation]): the time and beat numbers of bars and chord changes
```
Similarly, `(list, Observation)`.

genisplaja commented on mirdata/datasets/filosax.py:
```diff
@@ -0,0 +1,512 @@
+        self.annotation_path = self.get_path("annotations")
+
+    @property
+    def track_audio_property(self):
+        return "audio"
```
Don't worry about that one :)

genisplaja commented on mirdata/datasets/filosax.py:
```diff
@@ -0,0 +1,512 @@
+    @property
+    def segments(self):
+        """The times of segment changes (values are 'head', 'written solo', 'improvised solo')
+        Returns:
```
New line before `Returns:`!

genisplaja commented on mirdata/datasets/filosax.py:
```diff
@@ -0,0 +1,512 @@
+    @property
+    def bass_drums(self):
+        """The associated bass/drums track
+        Returns:
```
Same here and the rest :)

genisplaja reviewed (changes requested):
Hi @dave-foster! Just a small set of easy-to-address changes! The loader looks very good to me :) I'm just requesting small fixes, mainly related to code formatting and docs, and in my opinion, is ready to merge!!

Just a small thing. I think that if you click on [this link](https://mirdata--536.org.readthedocs.build/en/536/index.html) you'll be able to read the docs that are generated by this PR. Let me know if that doesn't work. Then you can go to `filosax` in `Dataset Loaders` in the menu and check that everything looks and reads nicely there!

Thanks again for your patience and timely responses. The loader looks really nice!! Let me know if you have further questions.

genisplaja commented on mirdata/datasets/filosax.py:
```diff
@@ -0,0 +1,512 @@
+
+@core.docstring_inherit(core.Dataset)
+class Dataset(core.Dataset):
+    """The Filosax dataset"""
```
If you take a look at the documentation link I attached at the end of the review, you can see that the documentation of the dataset class for `filosax` looks badly formatted.

genisplaja replied:
Sorry, @dave-foster! I think [this link](https://mirdata--536.org.readthedocs.build/en/536/source/mirdata.html) should work. In case it does not work, to see the docs rendered in your PR you need to get to [the doc most recent builds](https://readthedocs.org/projects/mirdata/builds/) and click on the latest build generated by your PR (#536), and then click on `View docs` in the top right side of the page. That will redirect you to the docs your PR is rendering.

dave-foster replied:
Thanks @genisplaja - all docs now looking okay

genisplaja commented:
Hello @dave-foster! Did you see my small review? Do you need help with this? Thanks!!

dave-foster commented on mirdata/datasets/filosax.py:
```diff
@@ -0,0 +1,512 @@
+(Lite - 558MB)
+https://zenodo.org/record/5643734#.YYLQ-i2l3UI
+
+Unzip the downloaded file to the folder /Users/<username>/mir_datasets/, and remove the version number from the folder:
```
I've added a few words clarifying that a user can define their own `data_home` upon initialisation.

dave-foster commented on mirdata/datasets/filosax.py:
```diff
@@ -0,0 +1,512 @@
+        note_path = "" if self.annotation_path == None else self.annotation_path
+        if note_path == "":
+            print("Note data only available for Sax tracks.")
+            return [Note({})]
```
This is slightly convoluted, due to attempts to fix the Exception error (which you found to be somewhat unrelated to this). So now, if the user tries to get note annotations from the backing track, it fails extremely elegantly - with an empty list of `Note` objects. Is this okay?

genisplaja replied:
Yes, sure! I think I remember now. Then, would it be possible to add a small test to make sure that `[Note({})]` is returned when the path is None?

Another thing I'd like to mention is that (and I'm sorry to detect that now), I think that it'd be good to use `load_annotation()` here, and mention in the docstring of `load_annotation()` that it can also be used to load the notes which are only available for the sax track. 
Then `notes()` would look like this:

```
@core.cached_property
def notes(self) -> Optional[List[Note]]:
    """ The track's note list - only for Sax files
    Returns:
        * [Note] - ordered list of Note objects
    """
    if not self.annotation_path:
        print("Note data only available for Sax tracks.")
        return [Note({})]
    else:        
        return load_annotation(self.load_annotation)
```

Does that sound reasonable to you? I understand why to use note_path but I think it is a little bit overkill and you could briefly explain that in the docstring without the need of duplicating code :) Let me know what you think!

At some point, I also think that the print here may not be needed, as long as the warning (`"Note data only available for Sax tracks."`) is clearly stated in the docstring. Maybe it'll be a bit annoying to get this message multiple times once you have learned that only sax tracks have note data.

Well, I'm sorry to leave here that piece of text @dave-foster... I hope that this helps and I'm open to discussing these last details before the loader is merged. Thanks for addressing the rest of my review, I went through the docs, and looks ready!!

dave-foster replied:
I've implemented your version of the code (which is far better than what was there), and added a test

dave-foster commented:
Hi @genisplaja, thanks for the nudge, and apologies for the delay getting your prescribed fixes incorporated into the code now. I've addressed all of the issues that you mention, and have added comments to those where I'm not quite sure how to proceed. Let me know what I need to do from this point? I can't tell you how much I appreciate your intervention to get this over the line!

dave-foster commented:
Thanks again @genisplaja - I think that's everything covered now!

genisplaja commented on mirdata/datasets/filosax.py:
```diff
@@ -0,0 +1,510 @@
+        segments (list, Observation): the time of segment changes
+        bass_drums (Track): the associated bass/drums track
+        piano_drums (Track): the associated piano/drums track
+        sax (list, Track)): a list of associated sax tracks
```
`sax (list, Track))` you do have here an extra`)`!

genisplaja commented:
I just left a tiny comment but looks ready to me. I'll ping @nkundiushuti (and whoever else is interested) to take a look and we can merge this. Thank you very much for your work and patience @dave-foster!

nkundiushuti commented on mirdata/datasets/filosax.py:
```diff
@@ -0,0 +1,510 @@
+"""
+
+
+class Note:
```
this is not the ideal way of implementing this but I am fine with it. we have a an annotation class and a NoteData class inside annotations https://github.com/mir-dataset-loaders/mirdata/blob/master/mirdata/annotations.py 
ideally this should have been inherited Annotation. I suggest going on with this PR and opening another issue regarding this (which can be solved in the future).

nkundiushuti commented:
thanks @dave-foster . I left some comment regarding the Note class which I think could have been implemented using the existing functionalities. however I am happy with it and I think it can be merged

dave-foster commented:
Brilliant, huge thanks @nkundiushuti and @genisplaja! I'll certainly have a look into conforming the `Note` annotations to the existing functionality. When do you anticipate the `pip install` version being updated (currently 0.3.6 Nov 6 2021)? Just so I can update my documentation and let users know? Thanks again!
